### PR TITLE
Write a "features" page, clean up existing "features.md"

### PR DESCRIPTION
**Pull request by rwols**
There are many concepts that we talk about here. But for a new user trying
this package, it may come across as confusing. Because terms like
"Signature Help" and "Code Actions" are not native to Sublime Text.

### CODE REVIEW

File: docs/src/concepts.md [hunk 1]
```diff
@@ -0,0 +1,123 @@
+# Concepts
+
+This package enhances existing concepts from Sublime Text and introduces new concepts not native to Sublime Text. This page provides an overview of the most important capabilities. The capabilities/concepts are accessible in different ways. Some are accessible via the Context Menu by right-clicking with your mouse, or via the top Menu Bar. Others are part of an existing workflow. Almost all capabilities can also be bound to a key with a key binding.
+
+## Auto Complete [cooperating]
+
+The auto-complete functionality of Sublime Text is well-known to any user. It provides word completions from the current buffer, and, since ST4, completions from other files. It presents the auto-complete widget in a synchronous fashion.
+
+The LSP package enhances the auto-complete functionality.
```
Comment (rchl) [inline docs/src/concepts.md:9]: A little more specific maybe?

```suggestion
The LSP package enhances the auto-complete list with results provided by the language server.
```

File: docs/src/concepts.md [hunk 2]
```diff
@@ -0,0 +1,123 @@
+# Concepts
+
+This package enhances existing concepts from Sublime Text and introduces new concepts not native to Sublime Text. This page provides an overview of the most important capabilities. The capabilities/concepts are accessible in different ways. Some are accessible via the Context Menu by right-clicking with your mouse, or via the top Menu Bar. Others are part of an existing workflow. Almost all capabilities can also be bound to a key with a key binding.
+
+## Auto Complete [cooperating]
```
Comment (rchl) [inline docs/src/concepts.md:5]: I wonder if those words in the square brackets are necessary or even that it's clear what those mean.

File: docs/src/concepts.md [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+
+The auto-complete functionality of Sublime Text is well-known to any user. It provides word completions from the current buffer, and, since ST4, completions from other files. It presents the auto-complete widget in a synchronous fashion.
+
+The LSP package enhances the auto-complete functionality.
+
+## Goto Definition [replacing]
+
+Sublime Text provides a "goto-def" feature by indexing the files in your project, and building a database out of the parsed files. The feature is accessible by clicking on Goto > Goto Definition. Sublime will attempt to jump to the definition of the word enclosing the caret. The files are parsed according to the `.sublime-syntax` associated to them. Entities which are assigned the `entity.name.*` scope are considered to be a "definition". Because a single `.sublime-syntax` file has no knowledge of the project structure, there may be multiple such "definitions".
```
Comment (rchl) [inline docs/src/concepts.md:13]: Why use the word `goto-def`? Nowhere in ST it's called using such abbreviation. Could use full words.

File: docs/src/concepts.md [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+
+It is also incorrect to call these "problems", as hints are not really problems.
+
+Sublime Text has no concept of diagnostics (nor lint results or problems), and hence does not provide an API endpoint to push diagnostics to the end-user. This package invented its own diagnostics presentation system.
+
+The SublimeLinter package provides similar functionality.
+
+## Formatting [new]
+
+Formatting is the art of computing a minimal set of white space text replacements. Formatting may be applied manually through a command invocation, or automatically when saving the file. Sublime Text has no concept of Formatting.
```
Comment (rchl) [inline docs/src/concepts.md:71]: > Formatting is the art of computing a minimal set of white space text replacements.

I'm not sure about that. The formatting is more about formatting the document in a certain way. Whether it's done using minimal set of replacements or by replacing the whole document is not really relevant here, I'd say.

File: docs/src/concepts.md [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+- Contextual: The links are close to the code they are representing.
+- Interspersed: The links are located throughout your source code.
+
+This package presents Code Lenses as a greenish clickable annotation positioned to the right of the viewport. Alternatively, they can be presented as phantoms.
+
+Sublime Text has no concept of Code Lenses.
+
+## Server Commands
+
+In Sublime Text you can bind any runnable command to a key or add it to various UI elements. Commands in Sublime Text are normally written in the Python plugin host. A language server may provide a runnable command as well. These kinds of commands are wrapped in an `lsp_execute` Sublime command that you can bind to a key.
```
Comment (rchl) [inline docs/src/concepts.md:109]: > Commands in Sublime Text are normally written in the Python plugin host.

Doesn't sound right. They are written in the editor.
Maybe: "Commands in Sublime Text are normally supplied by plugins or packages written in Python.

File: docs/src/concepts.md [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+
+## Server Settings
+
+Regular Sublime Text packages provide settings to adjust their behavior either via a `.sublime-settings` file that you can override in `Packages/User/package-name.sublime-settings`. A package may also allow customization through the `Packages/User/Preferences.sublime-settings` file and/or syntax-specific or project-specific settings.
+
+A language server may itself also expose settings that you can use to customize its behavior. For instance certain linter settings. This package allows customizing language server settings.
+
+## Server Initialization Options
+
+Initialization Options are like Server Settings, except they are static in the sense that they cannot be changed once the language server subprocess has started.
```
Comment (rchl) [inline docs/src/concepts.md:119]: The `Server Settings` could be a local link.

Maybe also add: "Those can be provided using the `initializationOptions` object within the client configuration" where "client configuration" could be a link to https://lsp.sublimetext.io/guides/client_configuration/.

File: docs/src/concepts.md (hunk 4 above)
Comment (rwols) [reply docs/src/concepts.md:67]: Well, it's still supposed to only be whitespace changes.

**Comment by jwortmann** [comment]
I guess this would be a good place to add a few illustrating GIFs. There are some at #1450, and I made a few more in the Mariana color scheme:

 __Auto Complete__

auto_complete

__Goto Definition__

goto_definition

__Find References__

find_references


__Expand Selection__

expand_selection

__Code Actions__

code_actions

__Document Highlights__ (description missing here in this PR)

document_highlight

**Comment by rwols** [comment]
Good idea. Dumping one for hover here. 
hoverfeature

**Comment by rwols** [comment]
goto symbol

gotosymbolfeature

**Comment by rchl** [comment]
I would think the the Features would be a more obvious place for mentioning the features with images. Think of new people who come to the project and look for stuff this package can do. I don't think the Concepts page will be an obvious place to look for.

**Comment by rchl** [comment]
Ideally we would have gifs all using the default theme and scheme since that would be consistent and most familiar. That's why I personally didn't like the gifs from the linked issue.

Also gifs are annoying in themselves for showing stuff like that since it's hard to know when they start and end.

**Comment by jwortmann** [comment]
So, rename "concepts.md" to "features.md" and put the things that are currently at features.md under "Keybord Shortcuts"? (or make another page with "Commands")

> Ideally we would have gifs all using the default theme and scheme since that would be consistent and most familiar. That's why I personally didn't like the gifs from the linked issue.

Yes, that's why I used Mariana (and the Adaptive theme, but I think that should be okay). I used a 750x400 pixel size for the GIFs, so that they hopefully don't get scaled too much on the website. I don't know if the docs website supports embedded videos as an alternative to GIFs?

**Comment by rwols** [comment]
I added the gifs as hyperlinks. Hopefully that satisfies everybody.

**Comment by jwortmann** [comment]
Videos should work via
```html
<video width="640" height="480" controls>
  <source src="movie.mp4" type="video/mp4">
</video>
```
or alternatively in Markdown syntax with https://github.com/soulless-viewer/mkdocs-video.

Not sure if the links to the GIFs on GitHub are permanent, maybe they should better be saved into its own directory and referenced by relative links?

**Comment by rwols** [comment]
I'm fairly sure those image links will work forever (until GitHub.com ceases to exist at least :) ). It's just some CDM with a huge backend.

**Comment by predragnikolic** [comment]
> Also gifs are annoying in themselves for showing stuff like that since it's hard to know where those start and end.

I would not agree with this one. 

IMO gifs are good to showcase features. See the [TypeScript ST plugin](https://github.com/microsoft/TypeScript-Sublime-Plugin#screenshots). I could easily conclude what the plugin does.

----

Suggestion 1
The goto definition, implementation, declaration can be grouped into one section.  

Suggestion 2
There could be less text on the features page.

Each feature can be described with one sentence on what the feature does.  (see the image below)
Additionally add a tip about that feature that is not immediately obvious, like the `f12` key for the `More` link. 

**Result**: Someone will faster conclude what a feature does.

127 0 0 1_8000_LSP_features_ (1)

127 0 0 1_8000_LSP_features_ (2)

Keeping with the one sentence description suggestion,You will avoid:
- writing about ST implementation details like "It presents the auto-complete widget in a synchronous fashion",
- or describing how ST [parsing works](https://github.com/sublimelsp/LSP/pull/1817/files#diff-4d8d6feecb61688373adc54566e11d5071aff07908d15bfb9f8a0f1b18a64b64R24). :)  


Suggestion 3 
Subprocesses, Server Initialization Options, Server Settings are not really features. I would remove the `Subprocesses` entirely (that is an implementation detail, and I do not thing that a user should care about that). But feel free to move the Server Initialization Options and Server settings to a more suitable place(I haven't tough where would be a more convenient place for them)

Suggestion 4
I would show the gifs on the features page for each feature. 

Requirements for the gifs:
- gifs should use the default theme,
- and they all should have the same resolution. (see what resolution looks the best for the and use it)

 We can enforce those rules on PR review when someone wants to add a gif to the features page.

**Comment by predragnikolic** [comment]
If you need ideas for short text description about features, you can see https://github.com/sublimelsp/LSP/blob/docs/index-page/docs/src/index.md#getting-started 

(Note to myself, I created the `docs/index-page` branch a long time ago, I should probably delete it once this PR is merged :) )

**Comment by rwols** [comment]
> Keeping with the one sentence description suggestion,You will avoid:

I disagree here. It's important for any user to understand how ST works. .sublime-syntaxes, its indexer, and its goto-def functionality are what makes ST ST. And it's important to note how that functionality is _replaced_, instead of _enhanced_.

**Comment by rwols** [comment]
@predragnikolic the suggestions sound OK. Let's merge this and have forward progress. Feel free to enhance the docs with further pull requests.

**Pull request merged**